pyomo/contrib/mindtpy/algorithm_base_class.py

```diff
@@ -0,0 +1,2272 @@
+            config.logger.info(
+                'NLP subproblem failed to converge within time limit.')
+            self.results.solver.termination_condition = tc.maxTimeLimit
+        elif solve_terminate_cond is tc.maxIterations:
+            config.logger.info(
+                'NLP subproblem failed to converge within iteration limit.')
+        else:
+            raise ValueError(
+                'MindtPy unable to handle MILP main termination condition '
+                'of %s. Solver message: %s' %
+                (solve_terminate_cond, results.solver.message))
+
+    ##################################################################################################################################################################################################################
+    # nlp_solve.py
+
+    def solve_subproblem(self, config):
+        """Solves the Fixed-NLP (with fixed integers).
+
+        This function sets up the 'fixed_nlp' by fixing binaries, sets continuous variables to their intial var values,
+        precomputes dual values, deactivates trivial constraints, and then solves NLP model.
+
+        Parameters
+        ----------
+        config : ConfigBlock
+            The specific configurations for MindtPy.
+
+        Returns
+        -------
+        fixed_nlp : Pyomo model
+            Integer-variable-fixed NLP model.
+        results : SolverResults
+            Results from solving the Fixed-NLP.
+        """
+        fixed_nlp = self.working_model.clone()
+        MindtPy = fixed_nlp.MindtPy_utils
+        self.nlp_iter += 1
+
+        # Set up NLP
+        TransformationFactory('core.fix_integer_vars').apply_to(fixed_nlp)
+
+        MindtPy.cuts.deactivate()
+        if config.calculate_dual_at_solution:
+            fixed_nlp.tmp_duals = ComponentMap()
+            # tmp_duals are the value of the dual variables stored before using deactivate trivial contraints
+            # The values of the duals are computed as follows: (Complementary Slackness)
+            #
+            # | constraint | c_geq | status at x1 | tmp_dual (violation) |
+            # |------------|-------|--------------|----------------------|
+            # | g(x) <= b  | -1    | g(x1) <= b   | 0                    |
+            # | g(x) <= b  | -1    | g(x1) > b    | g(x1) - b            |
+            # | g(x) >= b  | +1    | g(x1) >= b   | 0                    |
+            # | g(x) >= b  | +1    | g(x1) < b    | b - g(x1)            |
+            evaluation_error = False
+            for c in fixed_nlp.MindtPy_utils.constraint_list:
+                # We prefer to include the upper bound as the right hand side since we are
+                # considering c by default a (hopefully) convex function, which would make
+                # c >= lb a nonconvex inequality which we wouldn't like to add linearizations
+                # if we don't have to
+                rhs = value(c.upper) if c.has_ub() else value(c.lower)
```

Comment:
Will this really happen in Pyomo?

Reply:
@ZedongPeng there are two-sided inequalities in Pyomo. Emma is raising a valid concern.

Reply:
@ZedongPeng, yep, you can make a constraint with one either by giving a 3-tuple as the `expr` or using `inequality`.
So this
```
m.c = Constraint(expr=(1, m.x, 4))
```
is the same as this:
```
m.c = Constraint(expr=inequality(1, m.x, 4))
```

Reply:
If you don't want to address this in this PR, maybe just open a ticket, and you can do it later? I don't think it's a bug *introduced* here--I just noticed it here.

Reply:
In Pyomo, how do we define and calculate the duals for two-sided inequalities?

Reply:
Good question--it took me a minute to figure this out. I think that when you ask for `m.dual[m.two_sided_constraint]`, you get the dual with the largest magnitude. So, if one of the sides of the inequality is binding, you have that dual. You know the other one is zero because of complementary slackness. But I guess this means that you need to check the value of the constraint body compared to the right-hand sides to know what you have... 

Alternatively, you could preprocess your model and break two-sided inequalities into two separate constraints initially and not have to deal with this case inside of the algorithms.